File: android/app/src/main/kotlin/com/app/deenly/deenly/tajweed/MelFrontend.kt
package com.rnr.deenfocus.tajweed

import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.exp
import kotlin.math.floor
import kotlin.math.ln
import kotlin.math.sqrt

data class MelResult(val features: FloatArray, val time: Int)
data class MelBucketResult(val padded: FloatArray, val bucketT: Int)

/**
 * NeMo-compatible log-mel frontend (see tajweed-lab/web/asr/mel.py + COREML_CONTRACT.md).
 * Port of ios/Runner/Tajweed/MelFrontend.swift, but implemented with a plain radix-2
 * FFT instead of Accelerate/vDSP — no vDSP-specific scale factors here, this follows
 * the reference numpy `rfft` formulation directly (see mel.py).
 */
object MelFrontend {
    const val SAMPLE_RATE = 16_000
    const val N_FFT = 512
    const val HOP_LEN = 160
    const val N_MELS = 80
    const val PREEMPH = 0.97f
    const val LOG_EPS = 1e-5f

    private val BUCKETS = intArrayOf(80, 160, 320, 640, 1280, 2560, 4800)
    private val melFilterbank: Array<FloatArray> by lazy { makeMelFilterbank() }
    private val hannWindow: DoubleArray by lazy { makeHannWindow(N_FFT) }

    /** Returns mel features shaped as flat `[80 * T]` in row-major (mel, time), plus time `T`. */
    fun logMel(pcm: FloatArray): MelResult {
        if (pcm.size <= N_FFT) {
            throw TajweedNativeException(TajweedErrorCode.AUDIO_TOO_SHORT, "Audio too short for mel features.")
        }

        val wav = FloatArray(pcm.size)
        wav[0] = pcm[0]
        for (i in 1 until pcm.size) {
            wav[i] = pcm[i] - PREEMPH * pcm[i - 1]
        }

        val pad = N_FFT / 2
        val padded = FloatArray(wav.size + pad * 2)
        for (i in 0 until pad) {
            padded[pad - 1 - i] = wav[minOf(i + 1, wav.size - 1)]
            padded[pad + wav.size + i] = wav[maxOf(0, wav.size - 2 - i)]
        }
        System.arraycopy(wav, 0, padded, pad, wav.size)

        val nFrames = 1 + (padded.size - N_FFT) / HOP_LEN
        if (nFrames <= 0) {
            throw TajweedNativeException(TajweedErrorCode.AUDIO_TOO_SHORT, "No mel frames.")
        }

        val bins = N_FFT / 2 + 1
        val powerSpec = Array(nFrames) { DoubleArray(bins) }
        val re = DoubleArray(N_FFT)
        val im = DoubleArray(N_FFT)
        for (t in 0 until nFrames) {
            val start = t * HOP_LEN
            for (i in 0 until N_FFT) {
                re[i] = padded[start + i] * hannWindow[i]
                im[i] = 0.0
            }
            Fft.forward(re, im)
            for (k in 0 until bins) {
                powerSpec[t][k] = re[k] * re[k] + im[k] * im[k]
            }
        }

        val mel = FloatArray(N_MELS * nFrames)
        for (m in 0 until N_MELS) {
            val fbRow = melFilterbank[m]
            for (t in 0 until nFrames) {
                var sum = 0.0
                val ps = powerSpec[t]
                for (k in 0 until bins) {
                    sum += fbRow[k] * ps[k]
                }
                mel[m * nFrames + t] = ln(sum + LOG_EPS).toFloat()
            }
        }

        for (m in 0 until N_MELS) {
            val offset = m * nFrames
            var mean = 0.0
            for (t in 0 until nFrames) mean += mel[offset + t]
            mean /= nFrames
            var variance = 0.0
            for (t in 0 until nFrames) {
                val d = mel[offset + t] - mean
                variance += d * d
            }
            variance /= nFrames
            val invStd = 1.0 / (sqrt(variance) + 1e-5)
            for (t in 0 until nFrames) {
                mel[offset + t] = ((mel[offset + t] - mean) * invStd).toFloat()
            }
        }

        return MelResult(mel, nFrames)
    }

    /** Nearest supported inference bucket T80…T4800 (kept for CoreML/ONNX bucket parity, not a hard ORT requirement). */
    fun padToBucket(features: FloatArray, time: Int): MelBucketResult {
        val bucket = BUCKETS.firstOrNull { it >= time }
            ?: throw TajweedNativeException(TajweedErrorCode.AUDIO_TOO_LONG, "Audio exceeds 48s limit.")
        if (time == bucket) return MelBucketResult(features, bucket)
        val out = FloatArray(N_MELS * bucket)
        for (m in 0 until N_MELS) {
            System.arraycopy(features, m * time, out, m * bucket, time)
        }
        return MelBucketResult(out, bucket)
    }

    /**
     * Symmetric Hann window matching numpy's `np.hanning(N)` exactly (denominator N-1),
     * since the ONNX encoder was calibrated against tajweed-lab/web/asr/mel.py, which
     * this Android frontend must match bit-for-bit-equivalent (not the periodic/ANE-power-
     * normalized variant iOS's vDSP_HANN_NORM uses — that only "works" there because the
     * CMVN mean-subtraction step cancels a uniform per-frame scale factor; ONNX Runtime has
     * no such Accelerate convenience API, so we replicate the reference formula directly).
     */
    private fun makeHannWindow(n: Int): DoubleArray {
        val w = DoubleArray(n)
        for (i in 0 until n) {
            w[i] = 0.5 - 0.5 * cos(2.0 * PI * i / (n - 1))
        }
        return w
    }

    private fun makeMelFilterbank(): Array<FloatArray> {
        val sr = SAMPLE_RATE.toDouble()
        val melMax = 1127.0 * ln(1.0 + (sr / 2) / 700.0)
        val melPts = DoubleArray(N_MELS + 2) { melMax * it / (N_MELS + 1) }
        val hzPts = melPts.map { 700.0 * (exp(it / 1127.0) - 1.0) }
        val binPts = hzPts.map { floor((N_FFT + 1) * it / sr).toInt() }
        val bins = N_FFT / 2 + 1
        val fb = Array(N_MELS) { FloatArray(bins) }
        for (m in 1..N_MELS) {
            val left = binPts[m - 1]
            val center = binPts[m]
            val right = binPts[m + 1]
            if (center != left) {
                for (k in left until center) {
                    fb[m - 1][k] = ((k - left).toFloat() / (center - left))
                }
            }
            if (right != center) {
                for (k in center until right) {
                    fb[m - 1][k] = ((right - k).toFloat() / (right - center))
                }
            }
        }
        return fb
    }
}

/** Minimal iterative radix-2 Cooley-Tukey FFT, in place on parallel real/imag arrays. */
internal object Fft {
    fun forward(re: DoubleArray, im: DoubleArray) {
        val n = re.size
        require(n and (n - 1) == 0) { "FFT size must be a power of two, got $n" }

        var j = 0
        for (i in 1 until n) {
            var bit = n shr 1
            while (j and bit != 0) {
                j = j xor bit
                bit = bit shr 1
            }
            j = j or bit
            if (i < j) {
                val tr = re[i]; re[i] = re[j]; re[j] = tr
                val ti = im[i]; im[i] = im[j]; im[j] = ti
            }
        }

        var len = 2
        while (len <= n) {
            val ang = -2.0 * PI / len
            val wr = cos(ang)
            val wi = kotlin.math.sin(ang)
            var i = 0
            while (i < n) {
                var curWr = 1.0
                var curWi = 0.0
                for (k in 0 until len / 2) {
                    val uRe = re[i + k]
                    val uIm = im[i + k]
                    val vRe = re[i + k + len / 2] * curWr - im[i + k + len / 2] * curWi
                    val vIm = re[i + k + len / 2] * curWi + im[i + k + len / 2] * curWr
                    re[i + k] = uRe + vRe
                    im[i + k] = uIm + vIm
                    re[i + k + len / 2] = uRe - vRe
                    im[i + k + len / 2] = uIm - vIm
                    val nextWr = curWr * wr - curWi * wi
                    val nextWi = curWr * wi + curWi * wr
                    curWr = nextWr
                    curWi = nextWi
                }
                i += len
            }
            len = len shl 1
        }
    }
}

File: ios/Runner/Tajweed/MelFrontend.swift
import Accelerate
import Foundation

/// NeMo-compatible log-mel frontend (see tajweed-lab/web/asr/mel.py + COREML_CONTRACT.md).
enum MelFrontend {
  static let sampleRate = 16_000
  static let nFFT = 512
  static let hopLen = 160
  static let nMels = 80
  static let preemph: Float = 0.97
  static let logEps: Float = 1e-5

  private static let melFilterbank: [Float] = makeMelFilterbank()
  private static let hannWindow: [Float] = makeHannWindow(nFFT)

  /// Returns mel features shaped as flat `[80 * T]` in row-major (mel, time), plus time `T`.
  static func logMel(pcm: [Float]) throws -> (features: [Float], time: Int) {
    guard pcm.count > nFFT else {
      throw TajweedNativeError(TajweedErrorCode.audioTooShort, "Audio too short for mel features.")
    }

    var wav = pcm
    // Pre-emphasis
    if wav.count > 1 {
      var prev = wav[0]
      for i in 1..<wav.count {
        let x = wav[i]
        wav[i] = x - preemph * prev
        prev = x
      }
    }

    // Reflect pad nFFT/2 on both sides
    let pad = nFFT / 2
    var padded = [Float](repeating: 0, count: wav.count + pad * 2)
    for i in 0..<pad {
      padded[pad - 1 - i] = wav[min(i + 1, wav.count - 1)]
      padded[pad + wav.count + i] = wav[max(0, wav.count - 2 - i)]
    }
    for i in 0..<wav.count {
      padded[pad + i] = wav[i]
    }

    let nFrames = 1 + (padded.count - nFFT) / hopLen
    guard nFrames > 0 else {
      throw TajweedNativeError(TajweedErrorCode.audioTooShort, "No mel frames.")
    }

    let window = hannWindow

    var powerSpec = [Float](repeating: 0, count: nFrames * (nFFT / 2 + 1))
    var frame = [Float](repeating: 0, count: nFFT)
    var realp = [Float](repeating: 0, count: nFFT / 2)
    var imagp = [Float](repeating: 0, count: nFFT / 2)
    var magnitudes = [Float](repeating: 0, count: nFFT / 2 + 1)

    let log2n = vDSP_Length(log2(Float(nFFT)))
    guard let setup = vDSP_create_fftsetup(log2n, FFTRadix(kFFTRadix2)) else {
      throw TajweedNativeError(TajweedErrorCode.inferenceFailed, "FFT setup failed.")
    }
    defer { vDSP_destroy_fftsetup(setup) }

    for t in 0..<nFrames {
      let start = t * hopLen
      for i in 0..<nFFT {
        frame[i] = padded[start + i] * window[i]
      }
      // Pack for real FFT
      realp.withUnsafeMutableBufferPointer { rp in
        imagp.withUnsafeMutableBufferPointer { ip in
          var split = DSPSplitComplex(realp: rp.baseAddress!, imagp: ip.baseAddress!)
          frame.withUnsafeBufferPointer { fb in
            fb.baseAddress!.withMemoryRebound(to: DSPComplex.self, capacity: nFFT / 2) { complex in
              vDSP_ctoz(complex, 2, &split, 1, vDSP_Length(nFFT / 2))
            }
          }
          vDSP_fft_zrip(setup, &split, 1, log2n, FFTDirection(kFFTDirection_Forward))
          // Power spectrum
          magnitudes[0] = split.realp[0] * split.realp[0]
          magnitudes[nFFT / 2] = split.imagp[0] * split.imagp[0]
          for k in 1..<(nFFT / 2) {
            let re = split.realp[k]
            let im = split.imagp[k]
            magnitudes[k] = re * re + im * im
          }
        }
      }
      // Scale for real FFT convention (vDSP)
      var scale: Float = 1.0 / 4.0
      vDSP_vsmul(magnitudes, 1, &scale, &magnitudes, 1, vDSP_Length(nFFT / 2 + 1))
      for k in 0...(nFFT / 2) {
        powerSpec[t * (nFFT / 2 + 1) + k] = magnitudes[k]
      }
    }

    // Mel projection: (nMels, nFFT/2+1) @ (nFFT/2+1, T) → (nMels, T)
    var mel = [Float](repeating: 0, count: nMels * nFrames)
    let bins = nFFT / 2 + 1
    for m in 0..<nMels {
      for t in 0..<nFrames {
        var sum: Float = 0
        let fbRow = m * bins
        let psRow = t * bins
        for k in 0..<bins {
          sum += melFilterbank[fbRow + k] * powerSpec[psRow + k]
        }
        mel[m * nFrames + t] = logf(sum + logEps)
      }
    }

    // Per-bin mean/var normalize across time
    for m in 0..<nMels {
      let offset = m * nFrames
      var mean: Float = 0
      mel.withUnsafeBufferPointer { buf in
        vDSP_meanv(buf.baseAddress! + offset, 1, &mean, vDSP_Length(nFrames))
      }
      var negMean = -mean
      mel.withUnsafeMutableBufferPointer { buf in
        let row = buf.baseAddress! + offset
        vDSP_vsadd(row, 1, &negMean, row, 1, vDSP_Length(nFrames))
        var variance: Float = 0
        vDSP_measqv(row, 1, &variance, vDSP_Length(nFrames))
        var inv = 1.0 / (sqrtf(variance) + 1e-5)
        vDSP_vsmul(row, 1, &inv, row, 1, vDSP_Length(nFrames))
      }
    }

    return (mel, nFrames)
  }

  /// Default mel-time buckets for the official HF offline-ANE multifunction
  /// pack (`predict_T80…T4800`). Prefer the active manifest's `encoderBuckets`
  /// at runtime so alternate packs can ship different bucket lists without a
  /// Swift change — see `ModelStore.encoderApi()`.
  static let defaultOfficialBuckets = [80, 200, 400, 800, 1600, 2400, 4800]

  /// Nearest supported CoreML bucket ≥ `time` from `buckets` (ascending).
  static func padToBucket(
    _ features: [Float],
    time: Int,
    buckets: [Int] = MelFrontend.defaultOfficialBuckets
  ) throws -> (padded: [Float], bucketT: Int) {
    let sorted = buckets.sorted()
    guard !sorted.isEmpty else {
      throw TajweedNativeError(TajweedErrorCode.modelLoadFailed, "encoderBuckets is empty.")
    }
    guard let bucket = sorted.first(where: { $0 >= time }) else {
      throw TajweedNativeError(
        TajweedErrorCode.audioTooLong,
        "Audio exceeds \(sorted.last!)-frame CoreML limit."
      )
    }
    if time == bucket { return (features, bucket) }
    var out = [Float](repeating: 0, count: nMels * bucket)
    // Copy each mel-row prefix; trailing pad stays 0 (identical to nested loops).
    features.withUnsafeBufferPointer { src in
      out.withUnsafeMutableBufferPointer { dst in
        guard let s = src.baseAddress, let d = dst.baseAddress else { return }
        for m in 0..<nMels {
          d.advanced(by: m * bucket).update(from: s.advanced(by: m * time), count: time)
        }
      }
    }
    return (out, bucket)
  }

  /// Zero-pads `features` (row-major `[80 * time]`) up to a single fixed `T`,
  /// used by non-multifunction (e.g. DIY-exported) CoreML encoders that only
  /// support one static input shape. Unlike `padToBucket`, callers must also
  /// pass the true `time` as a separate `length` model input — the DIY encoder
  /// keeps `length` as an explicit graph input (unlike the official multifunction
  /// packages, whose per-bucket functions bake the length in).
  static func padToFixed(_ features: [Float], time: Int, fixedT: Int) throws -> [Float] {
    guard time <= fixedT else {
      throw TajweedNativeError(
        TajweedErrorCode.audioTooLong, "Audio exceeds \(fixedT)-frame CoreML limit."
      )
    }
    if time == fixedT { return features }
    var out = [Float](repeating: 0, count: nMels * fixedT)
    for m in 0..<nMels {
      for t in 0..<time {
        out[m * fixedT + t] = features[m * time + t]
      }
    }
    return out
  }

  /// Symmetric Hann window matching numpy's `np.hanning(N)` exactly (denominator N-1) —
  /// same formula as Android's `MelFrontend.kt` `makeHannWindow` and the Python reference
  /// (`tajweed-lab/web/asr/mel.py`, `np.hanning(N_FFT)`). Computed manually in `Double`
  /// precision (matching Android) rather than via `vDSP_hann_window`'s `vDSP_HANN_NORM`
  /// flag, which applies Accelerate's own normalization and produces a measurably
  /// different window shape (TD-009) — not just a per-frame constant scale, which is why
  /// per-bin CMVN normalization did not fully cancel the discrepancy on real audio.
  private static func makeHannWindow(_ n: Int) -> [Float] {
    var w = [Float](repeating: 0, count: n)
    for i in 0..<n {
      w[i] = Float(0.5 - 0.5 * cos(2.0 * Double.pi * Double(i) / Double(n - 1)))
    }
    return w
  }

  private static func makeMelFilterbank() -> [Float] {
    let sr = Float(sampleRate)
    let nFft = nFFT
    let nMelsLocal = nMels
    let melMax = 1127.0 * logf(1.0 + (sr / 2) / 700.0)
    var melPts = [Float](repeating: 0, count: nMelsLocal + 2)
    for i in 0..<melPts.count {
      melPts[i] = melMax * Float(i) / Float(nMelsLocal + 1)
    }
    var hzPts = melPts.map { 700.0 * (expf($0 / 1127.0) - 1.0) }
    var binPts = hzPts.map { Int(floor((Float(nFft + 1) * $0) / sr)) }
    var fb = [Float](repeating: 0, count: nMelsLocal * (nFft / 2 + 1))
    for m in 1...nMelsLocal {
      let left = binPts[m - 1]
      let center = binPts[m]
      let right = binPts[m + 1]
      if center != left {
        for k in left..<center {
          fb[(m - 1) * (nFft / 2 + 1) + k] = Float(k - left) / Float(center - left)
        }
      }
      if right != center {
        for k in center..<right {
          fb[(m - 1) * (nFft / 2 + 1) + k] = Float(right - k) / Float(right - center)
        }
      }
    }
    return fb
  }
}
